### PR TITLE
Make CropTarget serializable

### PR DESCRIPTION
**Pull request by eladalon1983**

### CODE REVIEW

**Review by jan-ivar — CHANGES_REQUESTED** [review]

File: index.html [hunk 1]
```diff
@@ -19,6 +19,15 @@ <h2>Abstract</h2>
       a First Public Working Draft.
     </section>
     <section id="conformance"></section>
+    <section id="terminology">
+      <h2>Terminology</h2>
+      <p>
+        The terms
+        <dfn class="fixme" data-cite="HTML/structured-data.html#serializable-objects"
```
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:26]: Isn't that the canonical class-name for definitions?

File: index.html [hunk 2]
```diff
@@ -147,12 +156,31 @@ <h3><dfn>CropTarget</dfn> Definition</h3>
...
+          [Exposed = Window, Serializable]
           interface CropTarget {
             // Intentionally empty; just an opaque identifier.
           };
         </pre>
         <dl data-link-for="CropTarget" data-dfn-for="CropTarget"></dl>
+
+        <p id="serializable-croptarget">
+          {{CropTarget}} objects are [= serializable objects =] [[!HTML]]. Their [= serialization
+          steps =], given <var>value</var> and <var>serialized</var>, are:
```
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:168]: "The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm."

File: index.html [hunk 3]
```diff
@@ -147,12 +156,31 @@ <h3><dfn>CropTarget</dfn> Definition</h3>
...
+        </p>
+        <ol>
+          <li>Make <var>serialized</var> a randomly generated, globally unique {{DOMString}}.</li>
```
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:171]: Internal slot makes sense to me. Could you suggest the exact wording?

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (youennf) [reply index.html:167]: What is the use case for forStorage being set to true?

**Comment by youennf** [comment]
Overall it seems fine.
I would go with forStorage=false as a first step.

Also, if serialising then deserialising, we should get an equivalent object (representing the same element) but not exactly the same object.
I wonder whether we could have CropTarget have an internal slot storing its element.
Then the serialisation and deserialisation steps would be about reading/setting the CropTarget internal slot.

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (annevk) [reply index.html:167]: Do you want this to be able to end up in IDB/pushState/some other things or not, essentially.

File: index.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (annevk) [reply index.html:170]: You cannot set _serialized_ to a value. You need to set _serialized_.[[Data]] or some such to a value. (Otherwise you also overwrite _serialized_.[[Type]] and break the caller. This should probably be documented better, maybe file an issue against whatwg/html?

I don't really have enough context here to provide further suggestions about what might need to be done.

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:167]: I want it to be possible for this to be stringified and sent over a REST call, as I know some Web applications that can cooperate either in a single-tab or multi-tab setting, and they use REST (and REST-like) for as much as they can, since that's already set up and works for them. I think it's a valid use-case and there's no reason to forbid it.

[the remark this answers is not in the document]
Comment (annevk) [reply index.html:167]: You mean through Fetch? Serializable doesn't help with that.

[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:167]: If Serializable forStorage=true does not allow conversion to a string, could you please advise what does?

[the remark this answers is not in the document]
Comment (annevk) [reply index.html:167]: If you want to convert a platform object to a string, you need to define a stringifier for it. That seems like a topic for a new issue though as #21 doesn't cover it.

[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html:167]: * I've added a slot for [[Data]].
* Indeed, stringification is a separate issue. If you happen to know the mechanism, I'd be happy to hear it. At any rate, we can do so as a follow-up.

File: index.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html]: Added [[Data]].

**Comment by eladalon1983** [comment]
> Overall it seems fine. I would go with forStorage=false as a first step.
> 
> Also, if serialising then deserialising, we should get an equivalent object (representing the same element) but not exactly the same object. I wonder whether we could have CropTarget have an internal slot storing its element. Then the serialisation and deserialisation steps would be about reading/setting the CropTarget internal slot.

Done with an internal slot for the serialized object, but not for CropTarget, where that would be a different topic. PTAL at the PR.

File: index.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (eladalon1983) [reply index.html]: @jan-ivar, the current text allows the user agent to determine its own implementation. A mapping is one possibility.

**Comment by youennf** [comment]
Looking at the new PR, I think it is missing the forStorage handling, something like:
- Add forStorage as parameter to the serialisation steps algorithm.
- As first step of the algorithm, throw an exception if forStorage is set to true.

I also think it would be preferable to move away from using UUIDs, which is currently a half-defined mechanism anyway.
Instead, we could do the following:
- A CropTarget has a weak reference to its element as a private slot, say cropTarget.[[Element]] (weak reference since we do not want to keep the element alive if crop targets are still in the air).
- When serialisation happens, set serialised.[[cropTargetElement]] to cropTarget.[[Element]].
- When deserialisation happens, create a CropTarget and set its [[Element]] to serialised.[[cropTargetElement]].

**Comment by eladalon1983** [comment]
> Looking at the new PR, I think it is missing the forStorage handling, something like:
> 
> * Add forStorage as parameter to the serialisation steps algorithm.
> * As first step of the algorithm, throw an exception if forStorage is set to true.

On the PR itself, I have quoted this part:
> The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm.

It's taken from [here](https://whatpr.org/html/6523/structured-data.html#serializable-objects).
I read it as "ignore this if it's not relevant." Seems irrelevant enough to me. Or why do you think it's important to specify throwing?

> I also think it would be preferable to move away from using UUIDs, which is currently a half-defined mechanism anyway. 

The PR does not mention UUIDs. (And that's intentional.)

> Instead, we could do the following:
> 
> * A CropTarget has a weak reference to its element

Please recall that the element is in another document. Potentially a cross-origin document.

**Comment by youennf** [comment]
I drafted a version of what I think could be done at https://github.com/youennf/mediacapture-region/commit/e2cb285eb861d226419068cf85d6c2cdf8c9081a



> Please recall that the element is in another document. Potentially a cross-origin document.

Internal slots are not exposed to JS so I think it is fine.

**Comment by jan-ivar** [comment]
> I drafted a version of what I think could be done at https://github.com/youennf/mediacapture-region/commit/e2cb285eb861d226419068cf85d6c2cdf8c9081a

With https://github.com/w3c/mediacapture-region/pull/32/files#r834656511 merged this LGTM, modulo s/Element/ElementHandle/

@eladalon1983 Do you agree? If so, let's get these PRs merged.

File: index.html [hunk 4]
```diff
@@ -147,12 +155,39 @@ <h3><dfn>CropTarget</dfn> Definition</h3>
...
+        </p>
+        <ol>
+          <li>
+            Compare <var>serialized</var>.[[\Data]] against a record of all {{CropTargets}} produced
+            and serialized thus far in the current top-level [=browsing context=], or in any of its
+            descendent [=browsing contexts=].
+            <ol>
+              <li>
+                If <var>serialized</var>.[[\Data]] was obtained by serializing previous
+                {{CropTarget}}, then set <var>value</var> to an equivalent {{CropTarget}}.
```
Comment (youennf) [inline index.html:185]: This equivalent has a fuzzy definition.
That is one reason why I think we should introduce an element reference slot in the spec.

File: index.html [hunk 5]
```diff
@@ -147,12 +155,39 @@ <h3><dfn>CropTarget</dfn> Definition</h3>
...
+              </li>
+              <li>Otherwise, set <var>value</var> to {{undefined}}.</li>
```
Comment (youennf) [inline index.html:187]: If the serialisation steps are done correctly, we should always successfully deserialise, which does not seem to be the case here if we set value to undefined.
Is the intention to serialise a CropTarget and deserialise it into undefined?
Does this mean serialising a CropTarget from top level page A to top level page B then back to top level page A would not work?

Introducing a direct element <-> CropTarget link, for instance by having CropTarget having an element reference seems a good approach to tighten things.

File: index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (jan-ivar) [reply index.html:167]: > > Looking at the new PR, I think it is missing the forStorage handling, something like:
> > 
> > * Add forStorage as parameter to the serialisation steps algorithm.
> > * As first step of the algorithm, throw an exception if forStorage is set to true.

I agree with @youennf here.

> On the PR itself, I have quoted this part:
> 
> > The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm.
> 
> It's taken from [here](https://whatpr.org/html/6523/structured-data.html#serializable-objects). I read it as "ignore this if it's not relevant."

I think you're misinterpreting it. It says _"The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm."_ which seems to say to not list the argument in the argument list unless it is referenced in the algorithm. Since the this algorithm would throw on it, that is satisfied.

Cc-ing @annevk as the authority on what the intent as far as default behavior here is, if any.

> Seems irrelevant enough to me. Or why do you think it's important to specify throwing?

We should narrow the scope of functionality to what's necessary for the use cases we're solving. If we don't need to implement storage, we shouldn't. The iterative process for adding functionality, is we hear use cases that demand it, and then discuss adding it.

It's easy to add more functionality later. What's hard is to remove it.

**Comment by jan-ivar** [comment]
> > Looking at the new PR, I think it is missing the forStorage handling, something like:
> > 
> > * Add forStorage as parameter to the serialisation steps algorithm.
> > * As first step of the algorithm, throw an exception if forStorage is set to true.

I agree with @youennf here.

> On the PR itself, I have quoted this part:
> 
> > The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm.
> 
> It's taken from [here](https://whatpr.org/html/6523/structured-data.html#serializable-objects). I read it as "ignore this if it's not relevant."

I think you're misinterpreting it. It says _"The introduction of these steps should omit mention of the forStorage argument if it is not relevant to the algorithm."_ which seems to say to not list the argument in the argument list unless it is referenced in the algorithm. Since the this algorithm would throw on it, that is satisfied.

Cc-ing @annevk as the authority on what the intent as far as default behavior here is, if any.

> Seems irrelevant enough to me. Or why do you think it's important to specify throwing?

We should narrow the scope of functionality to what's necessary for the use cases we're solving. If we don't need to implement storage, we shouldn't. The iterative process for adding functionality, is we hear use cases that demand it, and then discuss adding it.

It's easy to add more functionality later. What's hard is to remove it.

File: index.html (hunk 4 above)
Comment (eladalon1983) [reply index.html:185]: I think [this comment](https://github.com/w3c/mediacapture-region/pull/32#issuecomment-1080798232) applies here too; only its second half is relevant, so I'll repeat:
It might be that using weak references is standard procedure and would be well understood by readers. If that is the case, then I do not wish to delay us further. Could you point out an example of a Transferable/Serializable object that stores a weak reference to an Element in the original context, and where the weak reference gets replicated in the new context?

File: index.html (hunk 5 above)
Comment (eladalon1983) [reply index.html:187]: The `otherwise` section intended to deal with production of an invalid CropTarget. However, since we don't have a (legal) way to obtain an invalid CropTarget, I'd gladly drop this line - if that is indeed what you're suggesting...?

**Comment by eladalon1983** [comment]
> We should narrow the scope of functionality to what's necessary for the use cases we're solving. If we don't need to implement storage, we shouldn't. The iterative process for adding functionality, is we hear use cases that demand it, and then discuss adding it.
>
> It's easy to add more functionality later. What's hard is to remove it.

I suggest we document lack of consensus here and revisit this in the context of conversion to/from string. As I've mentioned to you out-of-band, I am in contact with Web-developers who often communicate using a shared cloud infrastructure (e.g. REST) even when communicating between frames in a given tab. For them, stringification of CropTargets is desirable.

**Comment by youennf** [comment]
stringification of CropTargets is orthogonal to serialisation of CropTargets.
forStorage is used as an example for IDB storage, this will not help your REST developers.

**Comment by eladalon1983** [comment]
> stringification of CropTargets is orthogonal to serialisation of CropTargets. forStorage is used as an example for IDB storage, this will not help your REST developers.

If you're sure about that, then we can throw on `forStorage` as far as I'm concerned.

**Comment by jan-ivar** [comment]
@eladalon1983 We need:
 1. To reference an element within the same process
 2. This reference must not prevent GC of the element in that process
 3. We need to pass this reference to another process _and back_
 4. We _don't_ need to dereference it in the other process 

Since we _won't_ dereference in other processes, I see nothing scary, unsafe, harder, permissive, or unclear.

> A capturer could then convert the string back to a CropTarget and use it to shave off pixels.

That's one level up, and seems irrelevant. I think the interesting part is _[how](https://w3c.github.io/mediacapture-region/#dom-browsercapturemediastreamtrack-cropto) "the user agent calculates the bounding box of the pixels belonging to the element, and crops the frame to the coordinates of this bounding box"_, which I would think has to happen in the capturee process where the element is, to track layout-movements of the element. Hence the _"and back"_.

**Comment by eladalon1983** [comment]
> We need to pass this reference to another process and back

Citation needed for "and back" - see below.

> which I would think has to happen in the capturee process where the element is

The process where the element originates does not need to be consulted again. That might be a valid implementation, but it's not a necessary implementation. Allow me to tell you about Chrome's implementation. Chrome "tags" the Element in the rendering pipeline as soon as `produceCropTarget()` is called. This tagging travels along to the GPU process, which collates data from the many processes associated with the different frames. The process where the Element lives never finds out that it's being captured, let alone that an active capture is being cropped to an Element in said process.

**Comment by jan-ivar** [comment]
> That proves that implementations are possible that do not involve a weak reference back to the Element.

I think this WG works off the idea that any implementation that is indistinguishable from a spec algorithm is a valid implementation.

> Things are more patently safe when such references are not posted cross-process.

Can you explain the specific threat vector you worry about? A "reference" can be an id or a key/string for the user agent to look something up later in a different context, a "tag" if you will. It doesn't need to mean a memory pointer. Even if it were implemented that way, it'd 'd be a pointer that would be unusable in other processes without some lookup mechanism invented by the user agent, since neither the element itself or its environment or any other information about it leaves or gets copied to any other process. The only thing that leaves the process is whatever we specify in the serialization steps, which will probably be some id or key.

While it's good to give implementation advice where it matters¹, specs normally don't care about these details, and trusts user agents to be responsible. Specs typically care mostly about JS observable behaviors, like GC, i.e. the [3 properties I mentioned](https://github.com/w3c/mediacapture-region/pull/24#issuecomment-1081142141).

My view is @youennf's PR satisfies these, and that the prose you've provided in response does not, but please illuminate me, so we don't keep tripping up over details like this.

---

<sub>1. In WebRTC we use a [[[KeyingMaterialHandle]]](https://w3c.github.io/webrtc-pc/#dom-rtccertificate-getfingerprints) internal slot for the arguably much more sensitive information. If this is unsafe, this would be good to learn.</sub>

**Comment by eladalon1983** [comment]
@jan-ivar, I am concerned that "weak reference" might not be a well understood term in the way it's used here. I am afraid readers would think "pointer" and be alarmed. If there is a history of using "weak reference" in the sense it's being used here, please do let me know. Otherwise, it seems prudent to keep to the beaten path here. All CropTarget aims to do is serve as input to cropTo(). I think we can find a formulation that highlights this fact.

**Comment by youennf** [comment]
> Another problem I have with a slot that holds a "weak reference" to an Element, is that it becomes needlessly scary when considering later extensions of this API to cropping a track derived from capturing another tab.

@eladalon1983, with regards to https://github.com/w3c/mediacapture-region/pull/24#issuecomment-1080970113, if you invoke security concerns, please provide as precise security assessments. I honestly do not understand your message, or how it weakens weak :0) references.

FWIW, the web is full of well dealt with relationships between objects from different origins, living in potentially different processes which have security implications. Security will be best ensured and assessed with precisely defined relationships, semantics  and algorithms
This is the intent of this work, there is no intent in changing the behavior or outlawing a particular implementation.

> If you're sure about that, then we can throw on `forStorage` as far as I'm concerned.

An important part of the standardisation is to cross check and make sure we are all on the same page after this work.
In other words, do not trust me and please cross-check ;o)

**Comment by jan-ivar** [comment]
>  If there is a history of using "weak reference" in the sense it's being used here, please do let me know. 

[HTML](https://html.spec.whatwg.org/#offscreencanvas-placeholder) uses _"An [OffscreenCanvas](https://html.spec.whatwg.org/#offscreencanvas) object may hold a weak reference to a placeholder canvas element,"_

**Comment by jan-ivar** [comment]
"Weak reference" is also the term the TAG uses in their design principles in [§ 5.3. Don’t expose garbage collection](https://w3ctag.github.io/design-principles/#js-gc).

**Comment by eladalon1983** [comment]
> Given GC should not be observable, whether weak or not is not a hard requirement, it is a clarification.

We're in agreement here.

> If you do not like weak

My concern is mostly with "reference".

> I am fine reverting to the original PR: no mention of weak but a GC note.

To ensure we're discussing the same thing, could you please link me to the revision you're referring to?

> @youennf: As of objects referencing other objects that are potentially in other process, this is an existing pattern that proved to work well, let's use it.
>
> @jan-ivar: [HTML](https://html.spec.whatwg.org/#offscreencanvas-placeholder) uses "An [OffscreenCanvas](https://html.spec.whatwg.org/#offscreencanvas) object may hold a weak reference to a placeholder canvas element,"

Thank you for the reference, @jan-ivar. Please allow me some time to dig into these more closely.

> "Weak reference" is also the term the TAG uses in their design principles in § 5.3. Don’t expose garbage collection.

That example seems to refer to "weak reference" precisely in the interpretation that I want to avoid, is it not? "This means that you shouldn’t expose any API that acts as a weak reference, e.g. with a property that becomes null once garbage collection runs." It is **this** interpretation of "weak reference" that I worry would be confusing.

**Comment by eladalon1983** [comment]
Btw, I realize that CropTarget could be posted over **from another tab** using a BroadcastChannel (even with Storage Partitioning). Moreover, that tab could be closed before deserialization occurs. I wonder if we should then:
* Fail deserialization?
* Set the internal slot to a "weak reference" to an Element that no longer exists?
* Set the internal slot to undefined?
(FWIW, this question would be trivially answered if we said that CropTarget is a UUID, or has a single internal slot that **is** a UUID. Or a unique string other than UUID, for that matter.)

**Comment by eladalon1983** [comment]
As a compromise, and in the interest of progress, let's go with Youenn's proposed PR, which he'll upload shortly.
